src/openms/include/OpenMS/KERNEL/DimMapper.h

```diff
@@ -0,0 +1,692 @@
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+// OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+// ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// --------------------------------------------------------------------------
+// $Maintainer: Chris Bielow $
+// $Authors: Chris Bielow $
+// --------------------------------------------------------------------------
+
+#pragma once
+
+#include <OpenMS/CONCEPT/Exception.h>
+#include <OpenMS/CONCEPT/CommonEnums.h>
+#include <OpenMS/DATASTRUCTURES/DPosition.h>
+#include <OpenMS/DATASTRUCTURES/DRange.h>
+#include <OpenMS/KERNEL/BaseFeature.h>
+#include <OpenMS/KERNEL/MSExperiment.h>
+#include <OpenMS/KERNEL/MSSpectrum.h>
+#include <OpenMS/KERNEL/Peak1D.h>
+#include <OpenMS/KERNEL/Peak2D.h>
+#include <OpenMS/KERNEL/RangeManager.h>
+
+#include <array>
+#include <memory>
+
+
+namespace OpenMS
+{
+  /**
+    @brief A base class for a dimension which represents a certain unit (e.g. RT or m/z).
+           Derived classes implement virtual functions, which receive a well-defined data type,
+           e.g. a Feature, and return the appropriate value for their dimension (the DimRT class would return the RT of the feature).
+           This makes it possible to extract dimensions using a runtime configuration of DimBase instances.
+           Very useful when mapping units (RT, m/z) to axis when plotting etc.
+
+           The reverse (X-Y coordinates to data type, e.g. Peak1D) is also possible using 'from...()' methods
+  */
+  class OPENMS_DLLAPI DimBase
+  {
+  public:
+    using ValueType = double;
+    using ValueTypes = std::vector<ValueType>;
+
+    /// No default c'tor
+    DimBase() = delete;
+
+    /// Custom c'tor with unit
+    DimBase(DIM_UNIT unit) :
+        unit_(unit) 
+    {}
+
+    /// Assignment operator
+    DimBase& operator=(const DimBase& rhs) = default;
+
+    /// Equality
+    bool operator==(const DimBase& rhs) const
+    {
+      return unit_ == rhs.unit_;
+    }
+
+    /// Copy derived objects to avoid slicing when dealing with pointers to DimBase
+    virtual std::unique_ptr<DimBase> clone() const = 0;
+
+    virtual ValueType map(const Peak1D& p) const = 0;
+    virtual ValueType map(const Peak2D& p) const = 0;
+    virtual ValueType map(const ChromatogramPeak& p) const = 0;
+    virtual ValueType map(MSExperiment::ConstAreaIterator it) const = 0;
+    
+    /// obtain vector of same length as @p spec; one element per peak
+    /// @throw Exception::InvalidRange if elements do not support the dimension
+    virtual ValueTypes map(const MSSpectrum& spec) const = 0;
+
+    /// obtain vector of same length as @p spec; one element per peak
+    /// @throw Exception::InvalidRange if elements do not support the dimension
+    virtual ValueTypes map(const MSChromatogram& chrom) const = 0;
+
+    virtual ValueType map(const BaseFeature& bf) const = 0;
+
+    virtual ValueType map(const PeptideIdentification& pi) const = 0;
```

Comment:
You could add these accessory functions, but they are probably meaningless for a lot of types, e.g. `PeptideIdentification::getIntensity()` ?? But you'd have to provide them just to make the GUI compile.
This 'decorator' approach I'm using here is nicer IMHO, since it keeps the interface of the types clean from unexpected things.